src/uct/sm/mm/mm_posix.c

```diff
@@ -79,14 +91,151 @@ static ucs_status_t uct_posix_test_mem(size_t length, int shm_fd)
     return status;
 }
 
+static size_t uct_posix_get_path_size(uct_pd_h pd)
+{
+    uct_mm_pd_t *mm_pd = ucs_derived_of(pd, uct_mm_pd_t);
+    uct_posix_pd_config_t *posix_config = ucs_derived_of(mm_pd->config,
+                                                         uct_posix_pd_config_t);
+
+    /* if shm_open is requested, the path to the backing file is /dev/shm
+     * by default. however, if shm_open isn't used, in case UCS_NO was set for
+     * use_shm_open or if UCS_TRY was set but using shm_open() was unsuccessful,
+     * the size of the path to the requested backing file is needed so that the
+     * user would know how much space to allocated for the rkey. */
+    if (posix_config->use_shm_open == UCS_YES) {
+        return 0;
+    } else {
+        return 1 + strlen(posix_config->path);
+    }
+}
+
+static ucs_status_t uct_posix_set_path(char *file_name, int use_shm_open,
+                                       const char *path, uint64_t uuid)
+{
+    ucs_status_t status;
+    int ret;
+
+    if (!use_shm_open) {
+        strncpy(file_name, path, NAME_MAX);
+    }
+
+    ret = snprintf(file_name + strlen(file_name), NAME_MAX - strlen(file_name),
+                   "/ucx_posix_mm_%s_%s_%016lx", ucs_get_user_name(),
+                   ucs_get_host_name(), uuid);
+    if ((ret >= NAME_MAX) || (ret < 1)) {
+        status = UCS_ERR_INVALID_PARAM;
+        return status;
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_posix_shm_open(const char *file_name, size_t length, int *shm_fd)
+{
+    ucs_status_t status;
+
+    /* Create shared memory object and set its size */
+    *shm_fd = shm_open(file_name, O_CREAT | O_RDWR | O_EXCL,
+                       UCT_MM_POSIX_SHM_OPEN_MODE);
+    if (*shm_fd == -1) {
+        ucs_error("Error returned from shm_open %m. File name is: %s",
+                  file_name);
+        status = UCS_ERR_SHMEM_SEGMENT;
+        goto err;
+    }
+    if (ftruncate(*shm_fd, length) == -1) {
+        ucs_error("Error returned from ftruncate %m");
+        status = UCS_ERR_SHMEM_SEGMENT;
+        goto err_shm_unlink;
+    }
+
+    return UCS_OK;
+
+err_shm_unlink:
+    close(*shm_fd);
+    if (shm_unlink(file_name) != 0) {
+        ucs_warn("unable to unlink the shared memory segment");
+    }
+err:
+    return status;
+}
+
+static ucs_status_t uct_posix_open(const char *file_name, size_t length, int *shm_fd)
+{
+    ucs_status_t status;
+
+    /* use open with the given path */
+    *shm_fd = open(file_name, O_CREAT | O_RDWR | O_EXCL, UCT_MM_POSIX_SHM_OPEN_MODE);
+    if (*shm_fd == -1) {
+        ucs_error("Error returned from open %m . File name is: %s", file_name);
+        status = UCS_ERR_SHMEM_SEGMENT;
+        goto err;
+    }
+
+    if (ftruncate(*shm_fd, length) == -1) {
```

Comment:
but then if it wasn't successful you need to know if you need to unlink or shm_unlink. (btw, i need to unlink in uct_posix_open).
also i wanted to have the ftruncate as part of the posix_shm_open and posix_open functions so that if you need to use it again you would have the 'full package' of calling it - both open and ftruncate.